HPOS Features: Revert to one feature

coreymckrill opened the pull request:
In #38993, the `custom_order_tables` feature was split into two separate features, one for choosing which tables to use, and one for toggling data synchronization. My understanding is that this was done in order to have both of those settings appear on the Features settings page, instead of having to create a separate page for other HPOS-specific settings. However, this added some extra complexity, which then led to #39250 causing a regression where extensions that declared compatibility with HPOS were no longer recognized as compatible.

### Changes proposed in this Pull Request:

With this PR, we're reverting back to only having one "feature" for HPOS, but introducing a way to add additional setting controls to the Features settings page. This also does some refactoring of the FeaturesController class to make it more "functional", so that logic for specific features can live closer to those features in the codebase, without having to add a bunch of feature-specific code to the class.

### How to test the changes in this Pull Request:

**Test the feature management for HPOS:**

1. In WP Admin on your test site, go to WooCommerce > Settings > Advanced > Features.
2. "Data storage for orders" should appear under "Experimental features".
3. Try switching between "WordPress post tables" and "High performance order storage" for the setting, and saving. After refreshing the screen, the value of the setting should be retained.
4. The same applies to checking/unchecking the box for "Keep the posts and orders tables in sync".
5. With HPOS activated and syncing turned **off**, create an order. Back on the Features screen the "Data storage" radio buttons should be disabled and there should be a message beneath the data syncing checkbox saying the orders must be synced before the tables can be switched.
6. Sync the orders. Back on the Features screen the settings should be changeable again.

**Test the general functionality of the Features controller:**

1. Try enabling and disabling other features on the Features screen.
2. Add this snippet to your site:
    ```php
	add_action(
		'woocommerce_feature_enabled_changed',
		function() {
			wp_die( 'Hook reached!' );
		}
	);
    ```
3. Now try enabling or disabling a feature. You should see a "Hook reached!" message, indicating that the `woocommerce_feature_enabled_changed` action is successfully firing after a feature status is changed.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Consolidate HPOS back into a single "feature" for the purposes of showing it on the Features settings screen.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

coreymckrill committed:
HPOS Feature: Revert to one official feature

In #38993, the `custom_order_tables` feature was split into two
separate features, one for choosing which tables to use, and one for
toggling data synchronization. My understanding is that this was done
in order to have both of those settings appear on the Features settings
page, instead of having to create a separate page for other
HPOS-specific settings. However, this added some extra complexity,
which then led to #39250 causing a regression where extensions that
declared compatibility with HPOS were no longer recognized as
compatible.

With this PR, we're reverting back to only having one "feature" for
HPOS, but introducing a way to add additional setting controls to the
Features settings page.

coreymckrill committed:
phpcs fixes

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

coreymckrill committed:
Ensure HPOS settings detected correctly during db validation

Fixes `test_verify_base_tables_adds_and_remove_notice` unit test.

vedanshujain committed:
Connect switching on the option with queuing sync.

vedanshujain reviewed (approved):
LGTM! I have added one more commit which links back the order synchronization scheduling when the switch for  data sync is flipped. Waiting on CI to pass before merging.

vedanshujain committed:
Link turning on feature with table creation.

vedanshujain committed:
Use transient instead of action for unit test.

coreymckrill committed:
Cleanup tweaks after a messy rebase

coreymckrill committed:
Fix features unit tests

coreymckrill committed:
Add hook for filtering feature definitions, update tests

coreymckrill committed:
Use new filter hook to add custom_order_tables feature definition

Also allow feature definition properties to be callbacks to that their
values can be determined dynamically right before render.

coreymckrill committed:
Move legacy status to feature definition

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

coreymckrill committed:
Move more hard-coded feature logic into the definition

coreymckrill committed:
Tweak handler for when sync option changes

coreymckrill committed:
Add missing @since in doc block

coreymckrill committed:
phpcs cleanup

coreymckrill commented:
@vedanshujain @jorgeatorres this is now ready for another look. It's a big changeset. If necessary I can try and figure out how to split it into multiple PRs, but it's pretty intertwined.

Key points:
* HPOS switches back to one feature definition.
* All of the logic for registering the HPOS feature is encapsulated in `CustomOrdersTableController`.
* The `FeaturesController` class has been refactored so that other future features can similarly be registered from the part of the code where they live rather than adding a bunch of additional logic in the features controller.
* Unit tests are updated to reflect this new approach.

coreymckrill committed:
Fix regression: HPOS not experimental now

Konamiman commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -491,11 +489,17 @@ public static function verify_base_tables( $modify_notice = true, $execute = fal
+				'data_sync' => FILTER_VALIDATE_BOOL,
+			)
+		);
+		if ( in_array( true, $hpos_settings, true ) ) {
```
Is this check really necessary now that HPOS is non-experimental and enabled by default on new installs? The tables should always exist, right?

coreymckrill replied:
🤔 I would guess that it's an extremely unlikely scenario, but possible, that this check might still be necessary because the install routine would run again if you disabled and then re-enabled the plugin.

Konamiman commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -172,22 +98,85 @@ public function __construct() {
+			 *
+			 * @param array $features The feature definitions.
+			 */
+			$this->features = apply_filters( 'woocommerce_feature_definitions', $features );
```
Do we really want to add this filter? I understand that the goal is to apply some separation of concerns, but the purpose of the features engine was to have a centralized place to handle WooCommerce _core_ features, without allowing plugins to make any changes, and that's why all the features are hardcoded; but adding this filter defeats the whole thing.

What if instead, the `CustomOrdersTableController->add_feature_definition` method is made public (the class is `Internal` anyway) and we just call it directly? This keeps the separation of concerns intact without having to add a new filter.

And I'd say that the same applies to for `woocommerce_feature_setting`, but I see that this one is already in trunk.

coreymckrill replied:
> the purpose of the features engine was to have a centralized place to handle WooCommerce core features, without allowing plugins to make any changes, and that's why all the features are hardcoded; but adding this filter defeats the whole thing.

🤔 Fair point. So instead of a filter, I think you mean we could have `FeaturesController->add_feature_definition` and call that in other classes to add features? It seems like that would work, as long as it didn't create any circular logic loops from classes each depending on each other, which is what happened in my initial attempt. I guess a further question when using the `add_feature_definition` method would be, what action hook should it be called on, to ensure that all the features get registered correctly?

> And I'd say that the same applies to for `woocommerce_feature_setting`

We could potentially still turn this into a method and then deprecate the filter.

I'll work on this and see what I can come up with. Thanks for your review!

Konamiman replied:
> you mean we could have FeaturesController->add_feature_definition and call that in other classes to add features?

No, I was thinking of doing it the other way around: other classes would have their own `get_feature_definition` method, and these would be called from inside `FeaturesController::get_feature_definitions`. This would be cleaner I think, since we would be able to see at a glance all the calls to these methods by just looking inside the `FeaturesController` class.

coreymckrill replied:
Ah, ok. I'm not as keen on that approach for a few reasons:

* Not as much "separation of concerns" since you'd still have to edit the FeaturesController file every time you want to add a new feature, and each feature class would then be a dependency of the FeaturesController
* Related, I'm leery of creating circular dependencies. In the case of `CustomOrdersTableController`, it already has the FeaturesController as an injected dependency for other reasons. If we then try to call a method of it from within the features controller itself, it gets tricky to not have an endless loop situation (as I mentioned elsewhere, I already struggled with this)
* Doesn't provide a standard interface for registering a feature, whereas using `FeaturesController->add_feature_definition` means every feature will register in the same way

> we would be able to see at a glance all the calls to these methods by just looking inside the FeaturesController class.

I understand the benefit of having one place to see which features are being registered, but I'm not sure it outweighs the benefits of keeping the FeaturesController uncoupled from the other parts of the codebase where the features live...

Konamiman replied:
> Not as much "separation of concerns" since you'd still have to edit the FeaturesController file every time you want to add a new feature, and each feature class would then be a dependency of the FeaturesController

Separation of concerns doesn't mean there are no dependencies, it only means that each class has a well defined purpose and responsibilites aren't spread across multiple classes (not arbitrarily, at least).

> I'm leery of creating circular dependencies. In the case of CustomOrdersTableController, it already has the FeaturesController as an injected dependency for other reasons.

Now that's a very legitimate concern (no pun intended). Very true, the `FeaturesController` is already referenced by other classes providing feature definitions; we could maybe workaround this by using static methods but we would be deviating from what we can consider clean code.

> I understand the benefit of having one place to see which features are being registered, but I'm not sure it outweighs the benefits of keeping the FeaturesController uncoupled from the other parts of the codebase where the features live

Fair enough. As long as the hook disappears and we clearly document `add_feature_definition` as the entry point for other parts of the code to define their own features, I'm happy.

coreymckrill replied:
Refactored to use a method instead of a filter in 453457ade5a58c2fd94e2a36d66b283660a3e252

Konamiman commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -401,54 +386,67 @@ private function register_post_type_for_order_placeholders(): void {
-		if ( ! $sync_complete ) {
-			$disabled_option = array( 'yes', 'no' );
+	private function get_hpos_setting_for_feature() {
+		if ( 'yes' === get_transient( 'wc_installing' ) ) {
```
Why is this check necessary?

coreymckrill replied:
I'm actually not sure. @vedanshujain this is from one of your commits, do you remember why this is here?

Konamiman commented:
![image](https://github.com/woocommerce/woocommerce/assets/937723/b5a7a7c7-e038-4716-a74a-4fec655fa9fd)

I see two problems here:

1. The tooltip popup and the message under the setting are redundant, only one is needed.
2. The "Manage orders syncrhonization" link in the popup is unreachable anyway, as the popup disappears as soon as the cursor goes away from the "?" icon (also, isn't the link a bit useless? Synchronization is handled from the features page, where we're at already) 🙂

Note also that there's an open pull request (https://github.com/woocommerce/woocommerce/pull/39988) to add a filter to allow switching datastores even when there are orders pending sync, including changing the styling and the wording of the message under the setting. Given that, I'd leave only that message, removing the popup.

coreymckrill committed:
Ensure compat with PHP < 8.0+

FILTER_VALIDATE_BOOL is an alias of FILTER_VALIDATE_BOOLEAN but it
wasn't introduced until 8.0

coreymckrill committed:
Switch feature registration to use a method instead of a filter

* Update CustomOrdersTableController to use the new method
* Update the unit tests

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

coreymckrill committed:
Revert change that's unnecessary after rebase

coreymckrill committed:
phpcs fixes

Konamiman commented:
@coreymckrill We're almost there, but now there are merge conflicts, sorry 😄

coreymckrill commented:
Yeah, it also appears that I broke a couple of e2e tests 😅 On my list for this week.

coreymckrill committed:
Remove the `option_key` prop from product_block_editor feature

This was not actually the correct option key for the feature, and it
was causing some e2e tests to fail (thankfully!)
